### PR TITLE
MultiSelect

### PR DESCRIPTION
**Pull request by Feichtmeier**
grafik

@jpnurmi any veto? :D

### CODE REVIEW

**Comment by Feichtmeier** [comment]
ehhh wait a sec, need to get the value out again :facepalm:

**Comment by Feichtmeier** [comment]
oook now it works :P

**Comment by jpnurmi** [comment]
Does it not look a bit old-fashioned? :older_man: What comes to the implementation, take a look at how CheckedPopupMenuItem is implemented. :)

**Comment by Feichtmeier** [comment]
Well, at least in gnome we had this idea to make those kind of check lists monochrome

What's wrong with the implementation?

**Comment by jpnurmi** [comment]
When implemented as a PopupMenuItem, it can be a normal checkable item that is not limited to multi-selection and doesn't need to be wrapped into a dummy PopupMenuItem.

**Comment by Feichtmeier** [comment]
Alright Mr. fancy pans :P 
WDYT of this instead?

[fancy.webm](https://user-images.githubusercontent.com/15329494/197240524-1e41e73d-4b05-4f70-9e14-e2afa465d24a.webm)

(I try to implement the interface you mentioned)

**Comment by jpnurmi** [comment]
Here's the Flutter implementation for benchmarking. Ideally, we'd only have some Yaru-prefixes but otherwise the same. :)
```dart
PopupMenuButton<MyEnum>(
  itemBuilder: (context) {
    return [
      for (final value in MyEnum.values)
        CheckedPopupMenuItem<MyEnum>(
          value: value,
          checked: enumSet.contains(value),
          child: Text(value.name),
        ),
    ];
  },
  onSelected: (value) {
    setState(() {
      if (enumSet.contains(value)) {
        enumSet.remove(value);
      } else {
        enumSet.add(value);
      }
    });
  },
)
```

**Comment by Feichtmeier** [comment]
https://user-images.githubusercontent.com/15329494/197246020-4575d292-6cf2-4444-b10b-c99b3b2b2bad.mp4

The material component has the advantage of just being there but our implementation has the advantage of letting the popup stay open which is becomes much better if you think of many filters like for example packagekitfilter

**Comment by Feichtmeier** [comment]
@jpnurmi there is one more problem
Since it is now the type of an enum and not of a set<enum> the value turns a tile to be highlighted
grafik
:S I'd say let's go with the previous approach

**Comment by Feichtmeier** [comment]
Also the thing leaks into the button, where the not-super-fluttery approach does not
grafik

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 1]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+            ),
+      title: DefaultTextStyle(
+        style: TextStyle(color: textColor, fontSize: 16),
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+class YaruCheckMenuItem<T> extends PopupMenuItem<T> {
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:185]: Could we follow the same naming scheme as in Flutter, please? It's immensely valuable to call the same things by the same names. It makes it easier for Flutter developers to find what they're looking for and maps better when using "adaptive platform" packages. Last but not least, Flutter is about to introduce menubars with nested submenus and there will be several new "Menu" -related types that are not to be mixed with PopupMenuButton and its PopupMenuItems.

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 2]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+  Widget? get child => super.child;
+
+  @override
+  PopupMenuItemState<T, YaruCheckMenuItem<T>> createState() =>
+      _CheckedPopupMenuItemState<T>();
+}
+
+class _CheckedPopupMenuItemState<T>
+    extends PopupMenuItemState<T, YaruCheckMenuItem<T>>
+    with SingleTickerProviderStateMixin {
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:215]: `SingleTickerProviderStateMixin` can be removed if we're not dealing with raw `AnimationControllers`

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 3]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+  @override
+  Widget buildChild() {
+    final borderColor = Theme.of(context).colorScheme.onSurface.withOpacity(
+          Theme.of(context).brightness == Brightness.light ? 0.4 : 1,
+        );
+    return IgnorePointer(
+      child: ListTile(
+        enabled: widget.enabled,
+        leading: widget.checked
+            ? AnimatedContainer(
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:230]: could you make an internal helper widget out of this check indicator? we can share the indicator with the basic checkable menu item and the multi-selection variant

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 4]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+      _CheckedPopupMenuItemState<T>();
+}
+
+class _CheckedPopupMenuItemState<T>
+    extends PopupMenuItemState<T, YaruCheckMenuItem<T>>
+    with SingleTickerProviderStateMixin {
+  @override
+  void handleTap() {
+    super.handleTap();
+  }
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:219]: we can remove this from here, let's override it in the specialized popup menu item for multi-selection

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 5]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+    super.padding,
+    super.height,
+    super.mouseCursor,
+    super.child,
+  });
+
+  final bool checked;
+
+  @override
+  Widget? get child => super.child;
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:206]: can be removed

File: lib/src/controls/yaru_popup_menu_button.dart [hunk 6]
```diff
@@ -102,3 +102,163 @@ class _YaruPopupDecoration extends StatelessWidget {
...
+
+  @override
+  PopupMenuItemState<T, YaruCheckMenuItem<T>> createState() =>
+      _CheckedPopupMenuItemState<T>();
+}
+
+class _CheckedPopupMenuItemState<T>
```
Comment (jpnurmi) [inline lib/src/controls/yaru_popup_menu_button.dart:213]: _YaruChecked... :)

**Review by jpnurmi — COMMENTED** [review]
i know multi-selection is the ultimate goal here but let's take it step by step and focus on the basic checkable popup menu item first. :) multi-selection is kind of a specialized variant of that. we can build it next once we have the basic checkable popup menu item to build upon.

**Comment by Feichtmeier** [comment]
I use a in-app widget for now https://github.com/ubuntu-flutter-community/software/pull/422

**Comment by Feichtmeier** [comment]
@jpnurmi implemented your change wishes

Can you tackle the do-not-close-on-selection-problem after this is merged?

**Comment by jpnurmi** [comment]
## YaruCheckedPopupMenuItem
```dart
class YaruCheckedPopupMenuItem<T> extends PopupMenuItem<T> {
  const YaruCheckedPopupMenuItem({
    super.key,
    super.value,
    this.checked = false,
    super.enabled,
    super.padding,
    super.height,
    super.mouseCursor,
    super.child,
  });

  final bool checked;

  @override
  PopupMenuItemState<T, YaruCheckedPopupMenuItem<T>> createState() =>
      _YaruCheckedPopupMenuItemState<T>();
}

class _YaruCheckedPopupMenuItemState<T>
    extends PopupMenuItemState<T, YaruCheckedPopupMenuItem<T>> {
  @override
  Widget buildChild() {
    return IgnorePointer(
      child: ListTile(
        enabled: widget.enabled,
        leading: _YaruCheckIndicator(checked: widget.checked),
        title: widget.child,
      ),
    );
  }
}
```
The usage is identical to Flutter:
```dart
YaruPopupMenuButton<MyEnum>(
  itemBuilder: (context) {
    return [
      YaruCheckedPopupMenuItem(
        ...
      ),
  },
  onSelected: (value) {
    ...
  },
)
```

[yaru_checked_popup_menu_item.webm](https://user-images.githubusercontent.com/140617/197379183-935d786a-6a6a-4669-983d-dc97e4a88288.webm)


## YaruMultiSelectPopupMenuItem
```dart
class YaruMultiSelectPopupMenuItem<T> extends PopupMenuItem<T> {
  const YaruMultiSelectPopupMenuItem({
    super.key,
    super.value,
    this.checked = false,
    super.enabled,
    super.padding,
    super.height,
    super.mouseCursor,
    super.child,
    this.onChanged,
  });

  final bool checked;
  final ValueChanged<bool>? onChanged;

  @override
  PopupMenuItemState<T, YaruMultiSelectPopupMenuItem<T>> createState() =>
      _YaruMultiSelectPopupMenuItemState<T>();
}

class _YaruMultiSelectPopupMenuItemState<T>
    extends PopupMenuItemState<T, YaruMultiSelectPopupMenuItem<T>> {
  var _checked = false;

  @override
  void handleTap() {
    setState(() => _checked = !_checked);
    widget.onChanged?.call(_checked);
  }

  @override
  void initState() {
    super.initState();
    _checked = widget.checked;
  }

  @override
  Widget buildChild() {
    return IgnorePointer(
      child: ListTile(
        enabled: widget.enabled,
        leading: _YaruCheckIndicator(checked: _checked),
        title: widget.child,
      ),
    );
  }
}
```

In this case, one is no longer selecting a single item so the handler needs to be per-item but something like this is as close as it gets :)

```dart
YaruPopupMenuButton<MyEnum>(
  itemBuilder: (context) {
    return [
      YaruMultiSelectPopupMenuItem(
        ...
        onChanged: (value) {
          ...
        },
      ),
  },
)
```

[yaru_multi_select_popup_menu_item.webm](https://user-images.githubusercontent.com/140617/197379251-8afac974-9dcd-4262-86e1-af50a8d31285.webm)

**Comment by jpnurmi** [comment]
Sorry I was busy most of yesterday but I was just looking into it last night. :)

**Comment by Feichtmeier** [comment]
np, yeah per-item on-tap was also how I did it in my custom solution
SHall I add your changes to this PR or in another PR?

**Comment by jpnurmi** [comment]
Is there anything we could do about the alignment? It looks somehow unbalanced :)

image

**Comment by jpnurmi** [comment]
what about `minLeadingWidth: 18`? (the size of the check indicator)

image

**Comment by Feichtmeier** [comment]
> minLeadingWidth: 18

done

what about this?

>  SHall I add your changes to this PR or in another PR?

**Review by jpnurmi — APPROVED** [review]
LGTM :+1:

**Comment by Feichtmeier** [comment]
@jpnurmi will you add the multi select thing? Or shall I copy your code?

**Comment by jpnurmi** [comment]
Sure, feel free to. 🙂 Does it work as expected?

**Pull request merged**